### PR TITLE
feat: run watchers

### LINKED ISSUE #12: Run watchers
Run watchers separated by minute, hour/day
And also separated by auth required/not required

### PR DESCRIPTION
**Pull request by negebauer**
- Separate watcher functions
- Keep all downloaded watchers at application root in `watchers`
- Validate watchers config, only run those that are valid
- Separate watchers in minute/hour or day and also auth/noAuth
- Run auth watchers using users auth info, save snapshots for each user
- Run noAuth watchers using global snapshot, save it

Closes #12

### CODE REVIEW

File: src/watchers/config.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  TEMP_DIR_PATH: `${__dirname}/tmp`,
+  DOWNLOADED_DIR_PATH: `${__dirname}/downloaded`,
```
Comment (negebauer) [inline src/watchers/config.js]: maybe move this to the top config?

File: src/watchers/download-watchers.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  fs.emptyDirSync(DOWNLOADED_DIR_PATH);
+  console.table(WATCHERS_LIST);
+  console.log();
+
+  await Promise.all(
+    WATCHERS_LIST.map(async (watcher, index) => {
+      const repoTempPath = `${TEMP_DIR_PATH}/${index}`;
+      const { url, branch, commit, watchers } = watcher;
+
+      if (!url) return console.warn(`WARN: No url for i:${index} ${url}`);
```
Comment (negebauer) [inline src/watchers/download-watchers.js]: move to `validate` function

File: src/watchers/validate-libs.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable no-console */
```
Comment (negebauer) [inline src/watchers/validate-libs.js]: This is how we're doing validation for now, this just asumes that we're actually passing the libs.

To make this better we should extend `@notify-watcher/core` to allow creation of an `execute` function that passes more libs than just the ones from `@notify-watcher/core`, this way we can validate the watchers lists against the created `execute` function.

File: src/watchers/run-watchers.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function stopRunning(id) {
+  MOCK_REDIS[id] = false;
+}
+
+function runWatchersAuth(watchers) {
+  watchers.forEach(async watcher => {
+    const { name: watcherName, watch } = watcher;
+    const users = Users.filter(user => user.subscriptions[watcherName]);
```
Comment (negebauer) [inline src/watchers/run-watchers.js]: extract to method (that one will take care of calling the actual db)

File: src/watchers/run-watchers.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    watcher: 'gtd',
+    snapshot: {},
+  },
+  {
+    watcher: 'vtr',
+    snapshot: {},
+  },
+];
+
+async function isRunning(id) {
```
Comment (negebauer) [inline src/watchers/run-watchers.js]: a lot of this functions will have to be changed after we have the `db` connection and models

File: src/watchers/run-watchers.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+async function stopRunning(id) {
+  MOCK_REDIS[id] = false;
+}
+
+async function usersForWatcher(watcherName) {
+  return Users.filter(user => user.subscriptions[watcherName]);
+}
+
+async function getUserWatcherAuth(user, watcherName) {
+  const subscription = user.subscriptions[watcherName];
```
Comment (sasalatart) [inline src/watchers/run-watchers.js]: `return user.subscriptions[watcherName].auth;`

Comment (negebauer) [reply src/watchers/run-watchers.js]: will be changed later anyway

File: index.js [hunk 7]
```diff
@@ -18,9 +18,12 @@ app.use(router.allowedMethods());
 const server = app.listen(config.PORT);
 
 async function startup() {
```
[the remark this answers is not in the document]
Comment (negebauer) [reply index.js]: will be changed after the api pr's are merged

**Pull request merged**